ADBDEV-7567: Fix lost update during inplace update.

KnightMurloc opened the pull request:
Fix lost update during inplace update.

These two commits address an issue where an inplace update could modify an already-dead tuple. This could occur when a transaction updates a row in pg_class or pg_database without acquiring a lock on the target relation. In such cases, a concurrent transaction performing an inplace update might update the old version of the tuple, and after the first transaction commits, that update would be lost.

Commit 809ae0f587dfb168477265590ef1a20bde4d4391 fixes this by acquiring a lock on tuples before updating them.

Most of the infrastructure required for this was introduced in commit b83e8fd701a9bc98a8eb41d8dd3811f2f0ec8fe5.

The tests were taken from commit 796d191028bb642bb04d9b3fd18d6f0aab7b1f1f and ported to isolation2, since in GPDB they require gdd to be enabled in order to pass.

---

Do not squash.

nmisch committed:
For inplace update durability, make heap_update() callers wait.

The previous commit fixed some ways of losing an inplace update.  It
remained possible to lose one when a backend working toward a
heap_update() copied a tuple into memory just before inplace update of
that tuple.  In catalogs eligible for inplace update, use LOCKTAG_TUPLE
to govern admission to the steps of copying an old tuple, modifying it,
and issuing heap_update().  This includes MERGE commands.  To avoid
changing most of the pg_class DDL, don't require LOCKTAG_TUPLE when
holding a relation lock sufficient to exclude inplace updaters.
Back-patch to v12 (all supported versions).  In v13 and v12, "UPDATE
pg_class" or "UPDATE pg_database" can still lose an inplace update.  The
v14+ UPDATE fix needs commit 86dc90056dfdbd9d1b891718d2e5614e3e432f35,
and it wasn't worth reimplementing that fix without such infrastructure.

Reviewed by Nitin Motiani and (in earlier versions) Heikki Linnakangas.

Discussion: https://postgr.es/m/20231027214946.79.nmisch@google.com

Differences from the original commit:

1. Most of the Assert code has been removed, as the corresponding infrastructure
is absent in 6X.

2. The definition of InplaceUpdateTupleLock has been moved to lock.h.

(cherry picked from commit cafcc3ad0ed324993211b0dce09362aad3165146)

nmisch committed:
Improve test coverage for changes to inplace-updated catalogs.

This covers both regular and inplace changes, since bugs arise at their
intersection.  Where marked, these witness extant bugs.  Back-patch to
v12 (all supported versions).

Reviewed (in an earlier version) by Robert Haas.

Discussion: https://postgr.es/m/20240512232923.aa.nmisch@google.com

Differences from the original commit:
1. Tests intra-grant-inplace and intra-grant-inplace-db are ported to
isolation2.

2. inplace-inval test adopted to 6X syntax.

(cherry picked from commit 796d191028bb642bb04d9b3fd18d6f0aab7b1f1f)

nmisch committed:
Fix data loss at inplace update after heap_update().

As previously-added tests demonstrated, heap_inplace_update() could
instead update an unrelated tuple of the same catalog.  It could lose
the update.  Losing relhasindex=t was a source of index corruption.
Inplace-updating commands like VACUUM will now wait for heap_update()
commands like GRANT TABLE and GRANT DATABASE.  That isn't ideal, but a
long-running GRANT already hurts VACUUM progress more just by keeping an
XID running.  The VACUUM will behave like a DELETE or UPDATE waiting for
the uncommitted change.

For implementation details, start at the systable_inplace_update_begin()
header comment and README.tuplock.  Back-patch to v12 (all supported
versions).  In back branches, retain a deprecated heap_inplace_update(),
for extensions.

Reported by Smolkin Grigory.  Reviewed by Nitin Motiani, (in earlier
versions) Heikki Linnakangas, and (in earlier versions) Alexander
Lakhin.

Discussion: https://postgr.es/m/CAMp+ueZQz3yDk7qg42hk6-9gxniYbp-=bG2mgqecErqR5gGGOA@mail.gmail.com

Differences from the original commit:

1. In the systable_inplace_update_begin function, the buffer for
heap_inplace_lock is taken from the SysScanDesc.

2. Since SysScanDesc does not contain the old tuple in 6X,
systable_inplace_update_begin now returns the old tuple.

3. For the same reason, the systable_inplace_update_finish and
systable_inplace_update_cancel functions now also accept the old tuple as a
argument.

4. In the index_update_stats function, the calculation of relpages for AO tables
is performed before the start of the inplace update. This is because the
functions GetAOTotalBytes and GetAOCSTotalBytes open the table, which leads to
access to pg_class, whose buffer is locked in systable_inplace_update_begin.
Therefore, the size calculation for AO tables is done prior to the inplace
update.

5. Updated comments.

6. Use the key as a single structure rather than an array to prevent potential
mistakes.

(cherry picked from commit 7354b680ab64fbde6072f248fe3b2ff909a99d12)

RekGRpth commented on src/backend/catalog/index.c:
```diff
@@ -2165,39 +2180,17 @@ index_update_stats(Relation rel,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relid));
+	systable_inplace_update_begin(pg_class, ClassOidIndexId, true, NULL,
+								  1, key, &oldtup, &state);
```
Can we store struct `TupleTableSlot *slot` in the `state` argument of the new functions?

KnightMurloc replied:
What advantage will this give us? In any case, we need `SysScanDesc` for `systable_endscan`.

red1452 commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,124 @@ systable_endscan_ordered(SysScanDesc sysscan)
+							  HeapTuple *p_oldtup,
+							  void **state)
+{
+	ScanKey		mutable_key = palloc(sizeof(ScanKeyData) * nkeys);
```
Can you add assertion, that `nkeys` must be more 0?

red1452 commented on src/backend/commands/vacuum.c:
```diff
@@ -1722,7 +1722,10 @@ vac_update_relstats(Relation relation,
 {
 	Oid			relid = RelationGetRelid(relation);
 	Relation	rd;
+	ScanKeyData key[1];
```
Why do we need array here (and other places, where array's size is 1)?

red1452 commented on src/backend/commands/vacuum.c:
```diff
@@ -2052,10 +2066,16 @@ vac_update_datfrozenxid(void)
+				ObjectIdGetDatum(MyDatabaseId));
+
+	systable_inplace_update_begin(relation, DatabaseOidIndexId, true,
+								  NULL, 1, key, &oldtup, &inplace_state);
```
You can use `sizeof(key)/sizeof(*key)` instead of 1 (one of advantages to use array). It can help to avoid mistakes in the future.

red1452 reviewed (commented):
> 3. The tests are taken from commit https://github.com/arenadata/gpdb/commit/796d191028bb642bb04d9b3fd18d6f0aab7b1f1f and ported to insolation2.

I think, it is bad idea. It is better to make a separate commit, because you get changes from another patch.

red1452 reviewed (commented):
Add "cherry-picked from ..." to the description of the last commit.

RekGRpth commented:
Where are the changes to the `dbcommands.c` file?

KnightMurloc commented:
> Where are the changes to the `dbcommands.c` file from first commit?

Because commit f18d3e47f0fe2743ea9d93b10fedd504ba818f4a is missing in 6X

> > Use the key as a single structure rather than an array to prevent potential
> > mistakes.
> 
> I think this is redundant because the `systable_inplace_update_begin` function takes as arguments the number of keys and an array of keys, not a pointer to a key.

Reverted

> Why weren't tests taken from the first and second commits?

Fixed

> Where are the changes to `catalog.c`, `indexcmds.c`, `execReplication.c`, `nodeModifyTable.c` from the second commit?

catalog.c - Because 6X lacks IsInplaceUpdateRelation.
indexcmds.c - Because 6X lacks update_relispartition
execReplication.c - Because 6X lacks execReplication.c and ExecSimpleRelationUpdate
nodeModifyTable.c - Because it was not possible to port the assert function because the 6X lacks the `LockHeldByMe` infrastructure.

> How about adding a new argument to the `systable_inplace_update_begin` function in the first commit to avoid copying outside of it?
> 
> ```c
> extern void systable_inplace_update_begin(Relation relation,
> 										  Oid indexId,
> 										  bool indexOK,
> 										  Snapshot snapshot,
> 										  int nkeys, const ScanKeyData *key,
> 										  HeapTuple *oldtup,
> 										  HeapTuple *oldtupcopy,
> 										  void **state);
> ```
> 
> This will avoid forgetting to copy.

Reworked

RekGRpth commented:
What does the `inplace-inval` test from the first commit check? Why does it pass without a patch?

RekGRpth commented:
Actualize the differences in the second commit description.

RekGRpth commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,127 @@ systable_endscan_ordered(SysScanDesc sysscan)
+	};
+
+	// The caller must make a copy of the tuple.
+	*p_oldtup = oldtup;
+	*state = scan;
+
+	return heap_copytuple(oldtup);
```
Maybe do it the other way around? Write the copy to the argument, as done in the original commit, and just return the original old tuple?

RekGRpth commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,123 @@ systable_endscan_ordered(SysScanDesc sysscan)
+ * places to use it.  Standard flow:
+ *
+ * ... [any slow preparation not requiring oldtup] ...
+ * systable_inplace_update_begin([...], &tup, &inplace_state);
```
```suggestion
* oldtup = systable_inplace_update_begin([...], &tup, &inplace_state);
```

RekGRpth commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,123 @@ systable_endscan_ordered(SysScanDesc sysscan)
+	Assert(snapshot == NULL);
+
+	Assert((relation->rd_id == RelationRelationId ||
+			relation->rd_id == DatabaseRelationId) || !IsSystemRelation(relation));
```
```suggestion
			relation->rd_id == DatabaseRelationId) ||
			!IsSystemRelation(relation));
```

RekGRpth commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,123 @@ systable_endscan_ordered(SysScanDesc sysscan)
+
+		Assert(scan->scan || scan->iscan);
+		if (heap_inplace_lock(scan->heap_rel,
+							  oldtup, scan->scan ? scan->scan->rs_cbuf : scan->iscan->xs_cbuf))
```
```suggestion
							  oldtup, scan->scan ?
							  scan->scan->rs_cbuf :
							  scan->iscan->xs_cbuf))
```

RekGRpth commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,123 @@ systable_endscan_ordered(SysScanDesc sysscan)
+	SysScanDesc scan = (SysScanDesc) state;
+	Relation	relation = scan->heap_rel;
+	Assert(scan->scan || scan->iscan);
+	Buffer		buffer = scan->scan ? scan->scan->rs_cbuf : scan->iscan->xs_cbuf;
```
pgindent this line

RekGRpth commented on src/backend/access/index/genam.c:
```diff
@@ -606,3 +609,123 @@ systable_endscan_ordered(SysScanDesc sysscan)
+	SysScanDesc scan = (SysScanDesc) state;
+	Relation	relation = scan->heap_rel;
+	Assert(scan->scan || scan->iscan);
+	Buffer		buffer = scan->scan ? scan->scan->rs_cbuf : scan->iscan->xs_cbuf;
```
pgindent this line

RekGRpth commented on src/backend/catalog/index.c:
```diff
@@ -2165,37 +2180,14 @@ index_update_stats(Relation rel,
+				ObjectIdAttributeNumber,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relid));
+	oldtup = systable_inplace_update_begin(pg_class, ClassOidIndexId, true, NULL,
+										  1, key, &tuple, &state);
```
pgindent these lines

RekGRpth commented on src/backend/catalog/toasting.c:
```diff
@@ -417,21 +419,36 @@ create_toast_table(Relation rel, Oid toastOid, Oid toastIndexOid,
+					ObjectIdAttributeNumber,
+					BTEqualStrategyNumber, F_OIDEQ,
+					ObjectIdGetDatum(relOid));
+		oldtup = systable_inplace_update_begin(class_rel, ClassOidIndexId, true,
+											   NULL, 1, key, &reltup, &state);
```
pgindent these lines

KnightMurloc replied:
What's wrong with this place? The line is within 80 characters.

RekGRpth replied:
okay

RekGRpth commented on src/backend/commands/vacuum.c:
```diff
@@ -1796,7 +1799,12 @@ vac_update_relstats(Relation relation,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relid));
+	oldtup = systable_inplace_update_begin(rd, ClassOidIndexId, true,
+										 NULL, 1, key, &ctup, &inplace_state);
```
pgindent these lines

RekGRpth commented on src/backend/commands/vacuum.c:
```diff
@@ -2052,8 +2065,13 @@ vac_update_datfrozenxid(void)
+				ObjectIdGetDatum(MyDatabaseId));
+
+	oldtup = systable_inplace_update_begin(relation, DatabaseOidIndexId, true,
+										  NULL, 1, key, &tuple, &inplace_state);
```
pgindent these lines

RekGRpth commented on src/backend/catalog/index.c:
```diff
@@ -2214,7 +2214,6 @@ index_update_stats(Relation rel,
 		BlockNumber relpages;
 		BlockNumber relallvisible;
 
-
```
?
